### ISSUE TITLE
Idea: contextual SystemParams

### ISSUE THREAD

**Issue by viridia** [bevyengine/bevy#11048]
## What problem does this solve or what need does it fill?

Currently `SystemParam` offers great flexibility in defining new kinds of injectable objects for systems. However, `SystemParam` is limited in one important respect, in that it can only derive results from `World`, and not from any narrower scope. That is, when we calculate the data that we want to inject, the only dynamic data we have to start with is `World`, and the only parameters we have for deciding what data gets injected are generic types. However, there are interesting use cases for a `SystemParam` that would have additional runtime context parameters.

Take for example, the case of an event handler such as use in `bevy_eventlistener`. The event object has several important pieces of context: the event itself, the target entity, and the listener entity (the object owning the handler callback). The callback function generally wants access to components and other objects that specifically relate to these entities, such as:

* A specific component on the target entity.
* An `EntityAccess` or `EntityMut` on the target entity.
* An `EntityCommands` on the target entity.
* A specific component on the listener entity.
* An `EntityAccess` or `EntityMut` on the listener entity.
* An `EntityCommands` on the listener entity.

Currently, the only way to access objects like this is to inject a query of all entities with the given component, and then have the handler function do the extra step of picking out from that query the one entity we're interested in. But it would be much cleaner if we could simply specify in the callback signature, "Hey, I need an `EntityMut` for the target entity only."

There are other possible use cases, but they all involve a one-shot system that is fired in the context of some entity that has special status for that system.

## What solution would you like?

To make this work, we'd need a special kind of one-shot system that accepted an additional parameter, which is some kind of context object. The individual `SystemParam`s would be able to read this context object when doing their injection magic, and use that to decide what data to inject. Both the `SystemParam` and the one-shot system would need to have matching types for the context data.

## What alternative(s) have you considered?

Everything that I described is doable with the existing functionality, but more verbosely and less elegantly.

## Additional context

It's an open question as to whether this is even possible, but it seems like it might be.

**Comment by pablo-lua** [comment]
Would that work as an FromWorld implementation? Like, where you can achieve the Param dynamically from the world?

**Comment by viridia** [comment]
@pablo-lua I'm not sure how it can, since there's no way to pass the runtime parameter to FromWorld.

Let me give an example of what I mean: suppose we're writing an event handler, and we want to be able to inject the debug name of the entity that received the event (this is presuming we are using something like bevy_eventlistener, where events can be targeted at a specific entity). So we could write something like:

```rust
fn handle_event(target_name: &TargetQuery<Name>) {
    println("The target of the event was: {}", target_name.get());
}
```
The `TargetQuery` is a contextual system param that knows about event targets. Unlike a regular query, which would return all `Name` components for all entities in the world, `TargetQuery` only returns the `Name` component for the targeted entity.

Without this contextual feature, what you have to write is something like:

```rust
fn handle_event(event: &Listener<Event>, name_query: &Query<Name>) {
    let target_name = name_query.get(event.target);
    println("The target of the event was: {:?}", target_name.get());
}
```
In other words, you have to look up all of the names of every entity in the world that has a name, and pick out the one for the target entity. The target entity is accessed via a different injected param (this is how bevy_eventlistener works).

**Comment by VVishion** [comment]
Since these contextual params dont work in regular systems you wouldn't want to implement `SystemParam` for them.

You'd either want to have a new trait like `HandlerSystemParam`, which will have a passtrough impl for all `SystemParam`s and special case your new params

or

You have a trait `HandlerSystem` impl for `FnMut(X, Y, Z, ..)` with all possible variations and whatever trait bounds or explicit types suiting the use case
e.g. `FnMut(Target, T0, T1, ...) where T0: SystemParam, ...`

or a combination of them.
But you can't rly generalize this, because "scope" is so arbitrary. This has to be implemented individually.

**Comment by VVishion** [comment]
Probably your imagined `&TargetQuery` would just do the same behind the scenes what you are currently doing. There is no benefit to mark a single entity as being referenced, because the ecs power lies in packed memory where you iterate over all entities with xy most of the time, anyway.

**Comment by alice-i-cecile** [comment]
Without a pressing use case in mind, I'm going to close this out as not-planned for now. There's quite a bit of design and implementation complexity here that I don't immediately see a good way to tackle.
